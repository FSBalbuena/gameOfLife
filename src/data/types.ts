export type cell = 0 | 1;
export type Grid = cell[][];
