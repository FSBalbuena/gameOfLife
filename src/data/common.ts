export const RANGE = 30;
export const EMPTY_GRID = Array.from({ length: RANGE }, () =>
  Array.from({ length: RANGE }, () => false)
);
