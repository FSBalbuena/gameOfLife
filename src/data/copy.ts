//---copy
export const SUCCESS = "success";
export const FAIL = "fail";
export const REQUIRED_GRID = "Board is required";
export const INVALID_GRID = "Board must be a 2D array of numbers";
export const INVALID_GRID_CELL = "Invalid Cell value";
export const REQUIRED_STEP = "Step is required";
export const INVALID_STEP_NUMBER = "Step must be a number";
export const INVALID_STEP_INT = "Step must be an integer";
export const INVALID_STEP_MIN = "Step must be a positive number";
export const STANDAR_ERR =
  "Sorry, something went wrong with next state calculation";

export const GAME_TITLE = "Conway's Game of Life";
export const GAME_DESCRIPTION = "A cellular automaton simulation.";
export const GAME_LOADING_DESCRIPTION = "Generating next state.";
export const GAME_NEXT_BUTTON = "Next state";
export const GAME_ADVANCE_BUTTON = "Advance state";
export const GAME_PLAY_FOREVER_CHECKBOX = "Play Forever";
