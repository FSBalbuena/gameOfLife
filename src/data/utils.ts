import { BoardErr, Grid, RequestErr } from "@/data/types";

export const createGridByRange = (range: number): Grid =>
  Array.from({ length: range }, () => Array.from({ length: range }, () => 0));

export const parseRequestErrToBoardErr = (error: RequestErr): BoardErr => ({
  steps: error?.steps?.[0],
  grid: error?.grid?.[0],
});
