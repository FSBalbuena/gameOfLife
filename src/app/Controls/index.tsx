"use client";
import {
  ChangeEventHandler,
  KeyboardEventHandler,
  useCallback,
  useState,
} from "react";
import { clsx } from "clsx";
import Image from "next/image";
import { stepsSchema } from "@/data/common";
import { GAME_ADVANCE_BUTTON, GAME_NEXT_BUTTON } from "@/data/copy";
import refreshIcon from "@/../public/restart-icon.svg";

type Props = {
  error?: string;
  disabled?: boolean;
  onGeneration: (steps?: number) => void;
  onRestart: () => void;
};
export default function Controls({
  error,
  onGeneration,
  disabled,
  onRestart,
}: Props) {
  const [steps, setSteps] = useState(1);

  const handleStepsChange: ChangeEventHandler<HTMLInputElement> = useCallback(
    (event) => {
      const value = Number(event.target.value);
      const validatedStep = stepsSchema.safeParse(value);
      if (validatedStep.success) {
        setSteps(value);
      }
    },
    []
  );

  const onAdvance = useCallback(
    () => onGeneration(steps),
    [steps, onGeneration]
  );

  const handleEnter: KeyboardEventHandler<HTMLInputElement> = useCallback(
    (event) => {
      if (event.key === "Enter") {
        onAdvance();
      }
    },
    [onAdvance]
  );

  return (
    <div>
      <div
        role="group"
        className={clsx(
          "flex-col gap-2 sm:flex-row sm:gap-10 flex w-full justify-center",
          {
            disabled,
          }
        )}
      >
        <button
          disabled={disabled}
          onClick={() => onGeneration()}
          type="button"
          className="fold-bold rounded border-2 border-black bg-white px-3 py-1 text-base font-bold text-black transition duration-100 hover:bg-yellow-400 hover:text-gray-900"
        >
          {GAME_NEXT_BUTTON}
        </button>

        <div className="flex">
          <input
            disabled={disabled}
            aria-describedby="steps-errors advance button"
            value={steps}
            onChange={handleStepsChange}
            onKeyDown={handleEnter}
            type="number"
            id="steps"
            className="border-2 p-1 rounded-md rounded-r-none w-10 text-center"
          />
          <button
            disabled={disabled}
            onClick={onAdvance}
            type="button"
            id="advance button"
            className="fold-bold rounded border-2 border-black bg-white px-3 py-1 text-base font-bold text-black transition duration-100 hover:bg-yellow-400 hover:text-gray-900"
          >
            {GAME_ADVANCE_BUTTON}
          </button>
        </div>
        <button
          disabled={disabled}
          onClick={onRestart}
          type="button"
          className="flex justify-center fold-bold rounded border-2 border-black bg-black px-1 py-1 text-base font-bold text-white transition duration-100 hover:bg-gray-900 hover:border-yellow-700"
        >
          <Image priority src={refreshIcon} alt="restart" className="w-6" />
        </button>
      </div>
      <p id="steps-errors" className="text-red-500 text-right m-4">
        {error}
      </p>
    </div>
  );
}
