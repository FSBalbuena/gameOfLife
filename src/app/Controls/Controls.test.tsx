import "@testing-library/jest-dom";
import { render, screen, fireEvent } from "@testing-library/react";
import Controls from "@/app/Controls";
import { GAME_NEXT_BUTTON, GAME_ADVANCE_BUTTON } from "@/data/copy";
const onGeneration = jest.fn();
const onRestart = jest.fn();
const disabled = false;
const error = "";
const defaultProps = {
  error,
  onGeneration,
  disabled,
  onRestart,
};

describe("Controls", () => {
  it("Renders without crashing", () => {
    const { container } = render(<Controls {...defaultProps} />);

    expect(container).toMatchSnapshot();
  });
  it("Renders disabled state", () => {
    const { container } = render(
      <Controls {...defaultProps} disabled={true} />
    );

    expect(container).toMatchSnapshot();
  });
  it("Renders without crashing error state", () => {
    const { container } = render(
      <Controls {...defaultProps} error="Something went wrong" />
    );

    expect(container).toMatchSnapshot();
  });

  describe("Interactive", () => {
    describe("When click on next button", () => {
      beforeEach(() => {
        onGeneration.mockClear();
      });
      afterEach(() => {
        onGeneration.mockClear();
      });
      it("it should call onGeneration", () => {
        render(<Controls {...defaultProps} />);
        const button = screen.getByRole("button", {
          name: GAME_NEXT_BUTTON,
        });
        expect(button).toBeInTheDocument();
        expect(onGeneration).not.toHaveBeenCalled();
        fireEvent.click(button as Element);
        expect(onGeneration).toHaveBeenCalled();
      });
    });

    describe("when uses Advance Steps", () => {
      beforeEach(() => {
        onGeneration.mockClear();
      });
      afterEach(() => {
        onGeneration.mockClear();
      });
      it('User should be able to type a number on "Advance Steps" input', () => {
        render(<Controls {...defaultProps} />);

        const input = screen.getByRole("spinbutton");
        expect(input).toBeInTheDocument();
        expect(input).toHaveValue(1);
        fireEvent.change(input, {
          target: { value: "2" },
        });
        expect(input).toHaveValue(2);
      });
      it('User should be able to click on "Advance steps" button.', () => {
        render(<Controls {...defaultProps} />);
        const input = screen.getByRole("spinbutton");
        expect(input).toBeInTheDocument();
        expect(input).toHaveValue(1);
        fireEvent.change(input, {
          target: { value: "2" },
        });
        expect(onGeneration).not.toHaveBeenCalled();
        const button = screen.getByRole("button", {
          name: GAME_ADVANCE_BUTTON,
        });
        expect(button).toBeInTheDocument();
        fireEvent.click(button);
        expect(onGeneration).toHaveBeenCalledWith(2);
      });
      it("User should be able to action with enter key.", () => {
        render(<Controls {...defaultProps} />);
        const input = screen.getByRole("spinbutton");
        expect(input).toBeInTheDocument();
        expect(input).toHaveValue(1);
        fireEvent.change(input, {
          target: { value: "2" },
        });
        expect(onGeneration).not.toHaveBeenCalled();
        fireEvent.keyDown(input, {
          key: "Enter",
        });
        expect(onGeneration).toHaveBeenCalledWith(2);
      });
      it("User should not be able to place a negative number, or zero.", () => {
        render(<Controls {...defaultProps} />);
        const input = screen.getByRole("spinbutton");
        expect(input).toBeInTheDocument();
        expect(input).toHaveValue(1);
        fireEvent.change(input, {
          target: { value: "-4" },
        });
        expect(input).toHaveValue(1);
        fireEvent.change(input, {
          target: { value: "0" },
        });
        expect(input).toHaveValue(1);
      });
    });
    describe("When click on restart button", () => {
      beforeEach(() => {
        onRestart.mockClear();
      });
      afterEach(() => {
        onRestart.mockClear();
      });
      it("it should call onRestart", () => {
        render(<Controls {...defaultProps} />);
        const button = screen.getAllByRole("button").at(2);
        expect(button).toBeInTheDocument();
        expect(onRestart).not.toHaveBeenCalled();
        fireEvent.click(button as Element);
        expect(onRestart).toHaveBeenCalled();
      });
    });
  });
});
