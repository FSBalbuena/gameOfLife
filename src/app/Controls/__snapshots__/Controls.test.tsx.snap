// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls Renders disabled state 1`] = `
<div>
  <div>
    <div
      class="flex-col gap-2 sm:flex-row sm:gap-10 flex w-full justify-center disabled"
      role="group"
    >
      <button
        class="fold-bold rounded border-2 border-black bg-white px-3 py-1 text-base font-bold text-black transition duration-100 hover:bg-yellow-400 hover:text-gray-900"
        disabled=""
        type="button"
      >
        Next state
      </button>
      <div
        class="flex"
      >
        <input
          aria-describedby="steps-errors advance button"
          class="border-2 p-1 rounded-md rounded-r-none w-10 text-center"
          disabled=""
          id="steps"
          type="number"
          value="1"
        />
        <button
          class="fold-bold rounded border-2 border-black bg-white px-3 py-1 text-base font-bold text-black transition duration-100 hover:bg-yellow-400 hover:text-gray-900"
          disabled=""
          id="advance button"
          type="button"
        >
          Advance state
        </button>
      </div>
      <button
        class="flex justify-center fold-bold rounded border-2 border-black bg-black px-1 py-1 text-base font-bold text-white transition duration-100 hover:bg-gray-900 hover:border-yellow-700"
        disabled=""
        type="button"
      >
        <img
          alt="restart"
          class="w-6"
          data-nimg="1"
          decoding="async"
          height="40"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
      </button>
    </div>
    <p
      class="text-red-500 text-right m-4"
      id="steps-errors"
    />
  </div>
</div>
`;

exports[`Controls Renders without crashing 1`] = `
<div>
  <div>
    <div
      class="flex-col gap-2 sm:flex-row sm:gap-10 flex w-full justify-center"
      role="group"
    >
      <button
        class="fold-bold rounded border-2 border-black bg-white px-3 py-1 text-base font-bold text-black transition duration-100 hover:bg-yellow-400 hover:text-gray-900"
        type="button"
      >
        Next state
      </button>
      <div
        class="flex"
      >
        <input
          aria-describedby="steps-errors advance button"
          class="border-2 p-1 rounded-md rounded-r-none w-10 text-center"
          id="steps"
          type="number"
          value="1"
        />
        <button
          class="fold-bold rounded border-2 border-black bg-white px-3 py-1 text-base font-bold text-black transition duration-100 hover:bg-yellow-400 hover:text-gray-900"
          id="advance button"
          type="button"
        >
          Advance state
        </button>
      </div>
      <button
        class="flex justify-center fold-bold rounded border-2 border-black bg-black px-1 py-1 text-base font-bold text-white transition duration-100 hover:bg-gray-900 hover:border-yellow-700"
        type="button"
      >
        <img
          alt="restart"
          class="w-6"
          data-nimg="1"
          decoding="async"
          height="40"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
      </button>
    </div>
    <p
      class="text-red-500 text-right m-4"
      id="steps-errors"
    />
  </div>
</div>
`;

exports[`Controls Renders without crashing error state 1`] = `
<div>
  <div>
    <div
      class="flex-col gap-2 sm:flex-row sm:gap-10 flex w-full justify-center"
      role="group"
    >
      <button
        class="fold-bold rounded border-2 border-black bg-white px-3 py-1 text-base font-bold text-black transition duration-100 hover:bg-yellow-400 hover:text-gray-900"
        type="button"
      >
        Next state
      </button>
      <div
        class="flex"
      >
        <input
          aria-describedby="steps-errors advance button"
          class="border-2 p-1 rounded-md rounded-r-none w-10 text-center"
          id="steps"
          type="number"
          value="1"
        />
        <button
          class="fold-bold rounded border-2 border-black bg-white px-3 py-1 text-base font-bold text-black transition duration-100 hover:bg-yellow-400 hover:text-gray-900"
          id="advance button"
          type="button"
        >
          Advance state
        </button>
      </div>
      <button
        class="flex justify-center fold-bold rounded border-2 border-black bg-black px-1 py-1 text-base font-bold text-white transition duration-100 hover:bg-gray-900 hover:border-yellow-700"
        type="button"
      >
        <img
          alt="restart"
          class="w-6"
          data-nimg="1"
          decoding="async"
          height="40"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
      </button>
    </div>
    <p
      class="text-red-500 text-right m-4"
      id="steps-errors"
    >
      Something went wrong
    </p>
  </div>
</div>
`;
